Cross-build http4s for Scala.js

armanbilge opened the pull request:
![](https://maven-badges.herokuapp.com/maven-central/com.armanbilge/http4s-core_2.12/badge.svg?style=flat)
**tl;dr** try it out with
```scala
val http4sV = "1.0-377-020cf9e"
libraryDependencies ++= Seq(
  // Node.js servers, not for serverless
  "com.armanbilge" %%% "http4s-ember-server" % http4sV,

  // Any Node.js environment, including servers, AWS Lambda, and CLI app
  "com.armanbilge" %%% "http4s-ember-client" % http4sV,

  // Any Node.js serverless environment, such as AWS Lambda or Google Functions
  "com.armanbilge" %%% "http4s-node-serverless" % http4sV

  // Browsers and also Cloudflare Workers
  "com.armanbilge" %%% "http4s-dom-fetch-client" % http4sV,

  // Browsers supporting service workers and also Cloudflare Workers
  "com.armanbilge" %%% "http4s-dom-service-worker" % http4sV,

  // any other modules from the list below :)
)
```

This PR introduces cross-builds to Scala.js for several of the core http4s modules per #956:
* `core`
* `laws`
* `testing`
* `tests`
* `server`
* `client`
* `ember-core`
* `ember-server`
* `ember-client`
* `node-serverless` **new!**
* `dom-core` **new!**
* `dom-fetch-client` **new!**
* `dom-service-worker` **new!**
* `dsl`
* `boopickle`
*  `jawn`
* `circe`

It's still a bit rough around the edges, but very nearly there I think (tests are passing on my machine). The main gaps for Scala.js are file- ~and crypto-related stuff~, not sure how important these are.

~At this time, there is no server or client implementation for Scala.js but I think that is best done in follow-up work :)~ I've implemented a fetch-based client for browsers, and ember is ~on the way via~ here! https://github.com/typelevel/fs2/pull/2453.

~Admin stuff: this is my first PR to this repo so that means **every** commit I push will have to be manually approved for CI, which makes it tricky to work independently. Could I possibly get manually whitelisted, or alternatively if there's another small PR I can get merged right away (e.g., fix typo etc.) that would also get me on the whitelist. Thanks!~

armanbilge commented:
All tests are passing!! Besides the occasional hiccup with `org.http4s.server.middleware.ThrottleSuite` which seems flaky on Firefox. I updated the test matrix following after CE3's, which tests the Scala.js modules on both Node.js and Firefox.

Now, I seek guidance on a couple remaining issues so we can hopefully get this merged :) Some TODO's I scattered in the code:

* [`CirceInstancesPlatform`](https://github.com/armanbilge/http4s/blob/js/circe/js/src/main/scala/org/http4s/circe/CirceInstancesPlatform.scala) for JS: the `CirceInstances` API offers several methods `jsonDecoderIncremental`, `jsonDecoderAdaptive`, `streamJsonArrayDecoder`, `jsonDecoderByteBufferImpl`. To keep parity across JVM and JS I implemented all these methods for JS, however they're just using Javascript's String-based JSON parser without doing anything optimizations as suggested by the method names. Ideas:

  1. Leave as-is.
  1. Drop API-parity, and offer only one (or two) honestly-named methods for JS.
  1. Try to improve the implementations to match the method names.

* Crypto for JS: `oauth1`, Websockets, `DigestUtil` all rely on crypto functions not available in Scala.js. Ideas:
  
  1. Drop these for JS.
  1. Implement with the [Web Crypto API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Crypto_API) also [experimentally available](https://nodejs.org/api/webcrypto.html#webcrypto_web_crypto_api) for Node.js.
  1. Create several implementations using JS-platform specific APIs for web, Node.js, etc. Provide these as separate modules or maybe try to select dynamically at runtime based on what's available.
  1. Find/roll-our-own crypto implementations in pure Scala??? 😬

A couple other more minor things:
* No file system stuff for JS, but we could support this with a separate module targeting Node.js.
* There's something strange with floats on `Scala.js` (try [`1.2f.toString`](https://scastie.scala-lang.org/hKPeJeVdSXeC4N9CAFO27w) and count the decimal places!) so I disabled these tests on JS.
* [DecodeSpec](https://github.com/armanbilge/http4s/blob/js/tests/shared/src/test/scala/org/http4s/DecodeSpec.scala) uses a bunch of weird charsets not available on JS, not sure if these are critical to the tests/can be rewritten.
* I think I messed up the sbt files by running `scalafmtSbt` and also a bunch of headers. Big deal?

Thanks in advance for your feedback/reviews!!

armanbilge commented:
After chatting with @ChristopherDavenport on Discord, my understanding is the thinking is to prioritize support for web-JS based users without locking out other JS platforms. So, I'm going to use the Web Crypto APIs to plug the crypto hole for now, and we can hopefully take advantage of its experimental presence in Node.js and/or provide a fallback implementation at a later point.

armanbilge commented:
Ugh, seems being able to test against Firefox is going to be non-trivial because the JS-build of munit assumes we are running on Node.js. CE3 gets around this issue by using specs2 🙃 Given that we are largely targeting browser-users I think a Firefox test suite is important. So either we switch to specs2 (or another compatible test lib) or try and convince munit to relax their requirements...

armanbilge commented:
I've now ported all the crypto stuff for core (websockets), client (oauth), and server (CSRF) to Scala.js via the WebCrypto API available in browsers and (experimentally) Node.js 15+.

Of note: I had to make my first public API-breaking changes. WebCrypto is fully asynchronous (promise-based) for just about everything, so I had to make several methods `F[_]: Async`-based. Some of them already were monadic, but required only a weaker typeclass.

armanbilge commented:
Following https://github.com/typelevel/keypool/pull/225 I believe we are now a hop-and-skip away from ember on Scala.js, that hop being crossing fs2.io.net to Scala.js.

cquiroz commented:
I'm so glad somebody is doing this. I had a previous scala.js port partially working but by now that's totally obsolete. Starting from scratch is the right move

armanbilge commented:
I've now implemented the first JS-based client `fetch-client` in 5817cca! It specifically targets browsers and wraps around the Fetch API. Unfortunately I haven't been able to test it yet for two reasons:
1. The current `ClientRouteTestBattery` only runs on JVM, and
2. Even though I can port it to Node.js, that doesn't help us test inside a browser, which is exactly what'd we want to be doing. In general, as browsers can't host servers I'm not quite sure how to do this (my working idea is to use a service worker to emulate a server). Frustratingly this is all academic until https://github.com/scalameta/munit/pull/376 merges so we can actually run tests in the browser.

Also, revisiting my earlier comment that

> my understanding is the thinking is to prioritize support for browser-based users without locking out other JS platforms

@djspiewak just informed me that Node.js-based users are just as important a target, e.g. for running on AWS lambda and other serverless frameworks. My goal is that https://github.com/typelevel/fs2/pull/2453 will unlock ember-server and ember-client for exactly this use case.

If you have political/social capital to spend, I would appreciate all the help I can get moving these upstream PRs along 😉

armanbilge commented:
I'm very happy to say we now have a working `fetch-client` for browsers! It passes the `ClientRouteTestBattery` suite running in Firefox via Selenium on my local; unfortunately CI seems to be struggling due to memory issues ...

armanbilge commented:
Remaining TODO, for the idealized merge (or can be completed in subsequent PRs).
Ordered from easiest to hardest.

- [ ] Test more modules in browser (so far, only `fetch-client`). Blocked by:
  1. https://github.com/scalameta/munit/pull/376

- [ ] Cross-build `jawn` module and re-work `circe` module (current implementation is minimally functional). Blocked by:
   1. https://github.com/typelevel/jawn/pull/351#issuecomment-872448976
   2. A forthcoming PR to circe once above is merged

- [ ] Re-work file-related stuff in `server` module. Blocked by:
  1. https://github.com/typelevel/fs2/pull/2453

- [ ] Cross-build `ember-{server,client}` for Scala.js. Blocked by:
  1. https://github.com/typelevel/fs2/pull/2453

Other upstream PRs that would be extremely beneficial:
- https://github.com/scala-js/scala-js-dom/pull/446
- https://github.com/ScalablyTyped/Converter/pull/202

ScalaWilliam commented:
This is fantastic! 👍

oyvindberg commented:
This is awesome, thank you! Also worked on this back in the day, but never got it far enough :)

ChristopherDavenport commented:
I got a basic project with a local release actually running ember-server locally. Requires to publishLocal from this branch.

https://github.com/ChristopherDavenport/js-test

armanbilge commented:
For everyone following along: we've now crossed the bulk of http4s to Scala.js, including ember client/server. I've published this branch, so please give it a try!

```scala
"com.armanbilge" %%% "http4s-<module-name>" % "1.0-148-8da8898"
```

See https://github.com/http4s/http4s/pull/4938#issue-677694130 for a list of all the modules available on Scala.js.

ScalaWilliam commented:
You guys are a-mazing.

Cannot believe http4s is going to be available on the JS platform. Even the servers.

ChristopherDavenport commented:
https://github.com/ChristopherDavenport/js-test/tree/browser - Full Server Backend Serving ScalaJS Dom on the fetch client example

ChristopherDavenport commented:
Note https://github.com/http4s/http4s/pull/4938/commits/d9f776a0b66e0ee0ca030dd5fa1076883ba7d7e2 passed CI green on first trial, and is being retried to see how stable the ci approach is.

ChristopherDavenport commented:
Green again, and with that there was consistent CI

armanbilge commented:
This merge was a bit scary, hopefully won't need to do again so extensively. Going to need some fixing to get green again.

rossabaker commented:
I've been ensconced in 0.23.0 and have only skimmed the above, but my thought is to release a 1.0 milestone (the first meaningfully different thing from 0.23) with this.  My assumption is that this is not going to be binary compatible?

armanbilge commented:
@rossabaker no worries at all, important stuff!! Actually I'm pretty sure this is completely binary compatible with the exception of some crypto APIs, see https://github.com/http4s/http4s/pull/4938#issuecomment-869108893.

rossabaker commented:
I guess we could try to run MiMa on it and see what else it finds.  But adding this to the 0.23 series is a big commitment if this is still experimental.  Did you have to make any significant compromises to retain compatibility?

I'll take care of the merge here, just to get oriented a bit.

armanbilge commented:
Oh wow, you want to bring it into 0.23 series? Woah 😅 

> Did you have to make any significant compromises to retain compatibility?

Not really. The only pain point that really sticks me (and continues to) is that the `Server` class uses JVM-only `InetSocketAddress` class instead of the cross-platform ip4s equivalent. This didn't cause a significant compromise per-say, it just means a lot of stuff that could have been a pure cross had to be split into JVM and JS implementations just to deal with this.

The trick to pulling this off was doing lots of cross-building of the downstreams. The extensive crossing in `fs2.io` is what makes the ember cross here nearly pure, and without compromise.

rossabaker commented:
Well, I was contemplating tradeoffs of 0.23, at least.  Since this diff is so massive already, it would be nice to keep 0.23 and main closer together, and it would let milestone-o-phobes try it sooner.  But it'll just make any merges from 0.22 to 0.23 that much harder.  I think keeping it on main for 1.0 is still right.

ip4s didn't exist when server was built on `InetSocketAddress`, and since it still compiled, it hasn't itched until now.  If it simplifies things, 1.0 is exactly the right time to make that kind of change.

armanbilge commented:
> it would be nice to keep 0.23 and main closer together, and it would let milestone-o-phobes try it sooner. But it'll just make any merges from 0.22 to 0.23 that much harder. I think keeping it on main for 1.0 is still right.

Makes sense! And sorry for creating additional complications. BTW before any milestone release with this stuff, it would be nice to slip in one more Node.js server implementation targeting lambda compute environments, ember isn't especially appropriate for these.

> If it simplifies things, 1.0 is exactly the right time to make that kind of change.

I thought so! Should I make the change in this PR, or maybe in a follow-up?

> Since this diff is so massive already

Also, the numbers +9,826 −1,916 are very misleading on this after accounting for all the generated headers (2x for every file split across platforms) and a `yarn.lock`. Still massive, but not _that_ massive :P

armanbilge commented on core/shared/src/main/scala/org/http4s/QValue.scala:
```diff
@@ -74,7 +74,7 @@ final class QValue private (val thousandths: Int) extends AnyVal with Ordered[QV
 }
 
-object QValue extends QValuePlatform {
+object QValue {
```
Oops?

armanbilge replied:
Oh these really aren't needed anymore? my bad :)

rossabaker replied:
No, those went away in 0.23.0.  They were in support of a long-deprecated macro, and we cleaned out the oldest deprecated things in #4998.

armanbilge replied:
Make sense, I had a momentary panic thinking it was one of the platform traits I introduced :)

armanbilge commented on core/shared/src/main/scala/org/http4s/Uri.scala:
```diff
@@ -172,7 +172,7 @@ final case class Uri(
 }
 
-object Uri extends UriPlatform {
+object Uri {
```
And again?

rossabaker commented:
> BTW before any milestone release with this stuff, it would be nice to slip in one more Node.js server implementation targeting lambda compute environments, ember isn't especially appropriate for these.

That's fine, but milestones are cheap.  I'd do whatever is an MVP, and we can keep refining in subsequent PRs.

> I thought so! Should I make the change in this PR, or maybe in a follow-up?

Again, whatever gets us an MVP that we can merge, but if it clears up complexity, this PR is good.  I haven't had time to study the PR in depth, so I don't know precisely which code we're talking about or workarounds that were necessary.  I'll trust your judgment.

> yarn.lock

Just as a warning, I'm probably going to update the shell.nix, because I haven't the faintest idea what it's going to take to build all this locally, but I want it to work on all my machines.  I'm probably also the only one who cares about the shell.nix. 😆

armanbilge commented:
>  I'm probably also the only one who cares about the shell.nix.

Lol, well hello shell.nix, didn't see you there ;) To _build_ this, I don't think you need anything more actually. But to _test_ it, you need a few things: Node.js 16, Firefox, geckodriver.

> Again, whatever gets us an MVP that we can merge

Since it sounds like that's what's important, I'd say this is exactly where we are at right now: we have a client for the browser and server/client for Node.js, all the test suites are passing in all relevant environments, and the CI is stable. @ChristopherDavenport can chime in too maybe since he's actually had a chance to play around with this.

ScalaWilliam commented on build.sbt:
```diff
@@ -500,6 +500,18 @@ lazy val jettyClient = libraryProject("jetty-client")
   )
   .dependsOn(core, testing % "test->test", client % "compile;test->test")
 
+lazy val nodeServerless = libraryProject("node-serverless", CrossType.Pure, List(JSPlatform))
```
what is this?? 

I am subscribed to your PR so getting notifications on every commit, this is incredible.

![tenor](https://user-images.githubusercontent.com/2464813/127752718-c9c754cd-6029-4fe0-9689-2794eea2a64c.gif)

armanbilge replied:
Yup, this is where this where we were headed all along :)

armanbilge commented:
Ok, I'd say that's a wrap :) I got in the last couple things I wanted to, there's something good for just about every JS environment.

@rossabaker I'm gonna make sure the CI is ✅ and then I'm ready to merge if you are!

armanbilge commented:
For everyone following along, I just published another snapshot with all the latest stuff. Details are in the top comment. https://github.com/http4s/http4s/pull/4938#issue-677694130

ChristopherDavenport commented:
This is green and good. I think we should expedite reviews so that way we don't end up with worse merge conflicts. @rossabaker  @cquiroz  @isomarcte

rossabaker commented on build.sbt:
```diff
@@ -111,95 +259,118 @@ lazy val core = libraryProject("core")
+    libraryDependencies += scalaJavaTime.value,
+    scalacOptions ~= { _.filterNot(_ == "-Xfatal-warnings") }
+  )
+  .jsConfigure(_.disablePlugins(DoctestPlugin))
```
:thinking: I was hoping to start leaning harder into these.  I guess most of the value is testing them on at least one platform.

armanbilge replied:
I was mostly getting into trouble with warnings for Scala.js facades, which have `js.native` definitions that made the compiler upset. I think I tried `@nowarn`ing them but then it was scala 2.12 vs 2.13, in the end this was a compromise. Happy to take another look.

rossabaker replied:
No, not a priority.  It only gets dicey if our examples become our only test coverage.  We're more likely to use it to ensure that our examples are correct.  Closer to an mdoc than a unit test.

cquiroz commented on core/shared/src/main/scala/org/http4s/util/Renderable.scala:
```diff
@@ -51,7 +50,7 @@ object Renderer {
       DateTimeFormatter
         .ofPattern("EEE, dd MMM yyyy HH:mm:ss zzz")
-        .withLocale(Locale.US)
+        // .withLocale(Locale.US) // Causes problems with Scala.js
```
What problems do you get?

armanbilge replied:
Ahh, I can fix this one now. While crossing Skunk I learned that the US locale is not included with Scala.js by default and it falls back to another one. To fix I can add a depenency to https://github.com/gemini-hlsw/gemini-locales/, do you mind looking at my PR there to cross-build it for 2.12 and 3?

rossabaker commented on build.sbt:
```diff
@@ -301,35 +489,90 @@ lazy val asyncHttpClient = libraryProject("async-http-client")
+    description := "Base library for dom http4s client and apps",
+    startYear := Some(2021),
+    libraryDependencies ++= Seq(
+      scalaJsDom.value.cross(CrossVersion.for3Use2_13),
```
`for3Use2_13` usually results in pestilence and death downstream.  Anywhere it's previously been in the build, we've used it speculatively compile the module for Scala 3 so we're ready when the dependency is, but not inflict diamond dependencies on the users.  Unless it's somehow different in the scala.JS world, I think we need to limit this to Scala 2.

armanbilge replied:
200% agree. I PRed the Scala 3 cross-build into that repo and then this happened https://github.com/scala-js/scala-js-dom/issues/451. The proper release is on the way soon, if we can bear with.

rossabaker replied:
:+1: As long as we remember not to publish it for Scala 3 until that happens, this is great.

rossabaker commented on build.sbt:
```diff
@@ -366,58 +609,63 @@ lazy val tomcatServer = libraryProject("tomcat-server")
       tomcatCatalina,
       tomcatCoyote,
-      tomcatUtilScan,
+      tomcatUtilScan
```
Our precious trailing commas. They are good for diff reduction!

armanbilge replied:
Must be the scalafmt config? 😆

rossabaker replied:
Probably.  We've deferred running it on the build until the branches settle down, because the merges are so bad.

cquiroz commented on ember-server/js/src/main/scala/org/http4s/ember/server/internal/WebSocketHelpersPlatform.scala:
```diff
@@ -0,0 +1,50 @@
+          Stream
+            .emit(value)
+            .through(fs2.text.encode(StandardCharsets.US_ASCII))
+            .++(Stream.chunk(Chunk.byteVector(ByteVector.view(Rfc6455.handshakeMagicBytes))))
```
Why `.++`?

armanbilge replied:
I'm sure I had a reason 😆 I assume you are asking, why the `.` syntax?

rossabaker commented on build.sbt:
```diff
@@ -429,12 +677,13 @@ lazy val scalaXml = libraryProject("scala-xml")
 
-lazy val twirl = http4sProject("twirl")
+// Full cross helps workaround issues with twirl directories
+lazy val twirl = http4sProject("twirl", CrossType.Full, List(JVMPlatform))
```
Is this going to publish as `http4s-twirl_2.13.6` instead of `http4s_twirl_2.13`?  I'd very much like to avoid that, if I understand it right.  What changed?

armanbilge replied:
Umm, yes? E.g. see https://search.maven.org/artifact/com.armanbilge/http4s-twirl_2.13/1.0-377-020cf9e/.

But also, that _is_ how _you_ publish it no?

https://search.maven.org/artifact/org.http4s/http4s-twirl_2.12

The `sbt-crossproject` plugin is really bad with "non-standard" source directories, e.g. see my PRs to add support for literally the basics. Doing a full cross seems to disappear these problems.

https://github.com/portable-scala/sbt-crossproject/pull/132
https://github.com/portable-scala/sbt-crossproject/pull/133

armanbilge replied:
BTW, we can totally revert this to the old way of doing things if we separate out `http4sProject(...)` and `http4sCrossProject(...)` for example. The nice thing about everything being a cross project is that you can programmatically identify which platforms each module supports, this is how I make the `rootJVM` and `rootJS for example.

rossabaker replied:
No, it doesn't look like it's doing what I was concerned about.  I thought it was putting the full Scala version in the artifact name instead of the binary version: _2.13.6 vs. 2.13.  That's not how we do things, but it's also not how your snapshot does things.  I misunderstand what that setting does, but it's not doing the wicked thing I thought it was, so I think it's fine.

rossabaker commented on build.sbt:
```diff
@@ -468,13 +717,14 @@ lazy val bench = http4sProject("bench")
 
-lazy val docs = http4sProject("docs")
+// Workaround via full cross
+lazy val docs = http4sProject("docs", CrossType.Full, List(JVMPlatform))
```
I don't think I care about this CrossType.Full, but I'd still like to understand it.

armanbilge replied:
Almost definitely the same issue as above. Or, even more weirdness :(

cquiroz reviewed (approved):
I'm so glad you have gotten this to work, great work. I wonder how big the resulting js would be mostly for use on the browser rather than in node, but at any rate this is great work

rossabaker commented on circe/src/main/scala/org/http4s/circe/middleware/JsonDebugErrorHandler.scala:
```diff
@@ -148,7 +148,7 @@ object JsonDebugErrorHandler {
           .asJson,
         "class_name" -> Option(a.getClass())
-          .flatMap(c => Option(c.getCanonicalName()))
+          .flatMap(c => Option(c.getName()))
```
Is canonical name not supported on scala.js?

armanbilge replied:
It's not, see https://github.com/scala-js/scala-js/issues/3452 If this is important we can split this file by platform :(

rossabaker replied:
It's, uh, usually the same.  As long as we call out the change.

rossabaker commented on client/js/src/main/scala/org/http4s/client/oauth1/oauth1platform.scala:
```diff
@@ -0,0 +1,71 @@
+import java.util.Base64
+import org.http4s.js.crypto.BufferSource
+
+private[oauth1] trait oauth1platform {
```
Thanks for making these private.  We didn't do this on our literal macros, and it made me sad earlier today.

rossabaker replied:
Nit: the name makes me twitch.  Can we capaitalize it like a regular trait: `Oauth1Platform`?

armanbilge replied:
Yup, easy fix :) tried my best to remember to mark platforms as private, let me know if I missed any!

rossabaker commented on client/jvm/src/main/scala/org/http4s/client/oauth1/oauth1platform.scala:
```diff
@@ -0,0 +1,42 @@
+
+private[oauth1] trait oauth1platform {
+  private val SHA1 = "HmacSHA1"
+  private def UTF_8 = StandardCharsets.UTF_8
```
Nit: any particular reason to not just import this?

armanbilge replied:
Nope, most likely I was following someone else's pattern here!

rossabaker replied:
I think that's right.

rossabaker commented on client/jvm/src/main/scala/org/http4s/client/oauth1/oauth1platform.scala:
```diff
@@ -0,0 +1,42 @@
+  private[oauth1] def makeSHASigPlatform[F[_]: Async](
+      baseString: String,
+      consumerSecret: String,
+      tokenSecret: Option[String]): F[String] = Async[F].pure {
```
Is this guaranteed not to fail?  Seems like an `F.catchNonFatal`, at least?

armanbilge replied:
Good point! Better be safe than sorry.

rossabaker commented on client/jvm/src/main/scala/org/http4s/client/oauth1/oauth1platform.scala:
```diff
@@ -0,0 +1,42 @@
+  private val SHA1 = "HmacSHA1"
+  private def UTF_8 = StandardCharsets.UTF_8
+
+  private[oauth1] def makeSHASigPlatform[F[_]: Async](
```
Isn't `Async` stronger than is needed here?  I think an `ApplicativeThrow` might do it.

armanbilge replied:
It is stronger, the `Async` matches what the JS side needs. I can dial this back.

armanbilge replied:
Ah, also relevant is your comment below about keeping APIs in sync.

rossabaker commented on client/shared/src/main/scala/org/http4s/client/oauth1/package.scala:
```diff
@@ -76,7 +72,7 @@ package object oauth1 {
       callback: Option[Callback] = None,
       verifier: Option[Verifier] = None
-  )(implicit F: MonadThrow[F], W: EntityDecoder[F, UrlForm]): F[Request[F]] =
+  )(implicit F: Async[F], W: EntityDecoder[F, UrlForm]): F[Request[F]] =
```
Thinking about what it means when a constraint is strengthened solely because one platform requires it.  I guess keeping the public APIs aligned is probably worth it.

armanbilge replied:
Yeah, this is a tricky one :(

rossabaker replied:
I'm a bit squeamish about "strongest constraint any platform needs."  If a future platform required stronger constraints following this rule, it's a breaking change to everybody.  But I can't imagine adding a third platform anytime soon, so this is an academic concern.

On the other hand, it's nice to present as uniform an interface as possible for things advertised as the same artifact.  Cross-builders on top of us would be in for a bad surprise if they build on `MonadThrow`, add scala.JS later, and find they need an `Async`.  Everything downstream would have to diverge in the same ways.

Yeah, I think you chose the lesser evil.

rossabaker commented on core/js/src/main/scala/org/http4s/js/crypto/Crypto.scala:
```diff
@@ -0,0 +1,937 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
Is this original work?  If not, can we attribute it and update NOTICES?  

If so, good grief, I'm sorry for the yaks you must shave.  Is this something that should be a library somewhere?  Or made `private[http4s]` so it's not an API liability for something that isn't actually directly HTTP?  That's what our internal package is about.

It is stupid o'clock here, and I am not qualified to review this package even at smart o'clock.  Is there anyone we can bring in?

armanbilge replied:
This is copy-pasta from https://github.com/scala-js/scala-js-dom/ (the `for3Use2_13` from that came up before :P). I'll fix the attribution/NOTICES.

Basicaly it's a really terrible situation where this identical API is available on Browsers and Node.js (see also https://github.com/http4s/http4s/pull/4938#discussion_r683102955) but the facades are written in such a way that they explode if they are on the wrong platform. You can read my story about this one here: https://users.scala-lang.org/t/scala-js-how-to-support-both-node-js-and-browser-platforms/7623

armanbilge replied:
Also, yes we should probably make the entire `js` package internal. At the time I was wondering if downstreams from http4s would want these utilities, but better not be on the hook as you say! Fortunately facades are cheap.

armanbilge replied:
Darn, we can't make just make this internal, I was being silly. At least some of these APIs _do_ need to be public (e.g., a `SecretKey` class for `CSRF`). I think that we're going to be on the hook for these after all :(

armanbilge commented on build.sbt:
```diff
@@ -2,29 +2,122 @@ import com.typesafe.tools.mima.core._
+    WorkflowStep.Use(
+      UseRef.Public("actions", "setup-node", "v2.1.5"),
+      name = Some("Setup NodeJS v16"),
+      params = Map("node-version" -> "16"),
```
Node 16 (experimentally) includes a `WebCrypto` API identical to the one in browsers. This let me quickly implement all the crypto methods in a cross-platform fashion. Node 16 will be the LTS in a few months I think.

armanbilge replied:
Also, it's worth mentioning: because JS imports these dynamically, people can still use Node 14 as long as they don't rely on any of these crypto APIs. TLS is still available either way, doesn't depend on Web Crypto.

armanbilge commented on build.sbt:
```diff
@@ -2,29 +2,122 @@ import com.typesafe.tools.mima.core._
+    cond = Some("matrix.ci == 'ciJVM'"))
+)
+
+val ciVariants = List("ciJVM", "ciNodeJS", /*"ciFirefox",*/ "ciChrome")
```
I finally gave up and removed it a few hours ago. The CI was really flaky on GH actions (works ok locally). Many of the CI failures between the last two green checks are related to that.

rossabaker commented on core/jvm/src/main/scala/org/http4s/RequestPlatform.scala:
```diff
@@ -0,0 +1,30 @@
+
+private[http4s] trait RequestPlatform[F[_]] { self: Request[F] =>
+
+  def remoteHost(implicit F: Sync[F]): F[Option[Hostname]] = {
```
What isn't supported here on scala.js?  Going through java.net?

armanbilge replied:
Yup, that would be it!

rossabaker commented on core/shared/src/main/scala/org/http4s/CacheDirective.scala:
```diff
@@ -1,11 +1,17 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
```
We lost a credit here (and in a few other places).  There's a list of files not to touch with generated headers in Http4sPlugin.scala.

armanbilge replied:
Yes, I know, sorry about it :( it got really messed up with the directory shuffling, despite my attempt to adjust. I can try again. Is there a way we can just have both headers, so the auto-generator can be happy?

rossabaker replied:
I never figured out a way.  sbt-header wants to control everything above the package line.  I guess attribution could appear somewhere below, like on the types.  Honestly, most of these files have changed enough that they barely resemble the originals, but I still try hard to respect original authors and their licenses.  We're not Microsoft.

My solution was to enumerate the files not to touch in a list in the build, which is fine until files start to move.

armanbilge replied:
I think I fixed all of these. `sbt-header` seems to be a little more friendly nowadays, as long as its on top.

rossabaker commented on core/shared/src/main/scala/org/http4s/Http4s.scala:
```diff
@@ -0,0 +1,35 @@
+package org.http4s
+
+@deprecated("Use org.http4s.implicits._ instead", "0.20.0-M2")
+trait Http4s extends Http4sInstances with Http4sFunctions with syntax.AllSyntax
```
I think we removed this entire file in 0.23 and it can be dropped here.

armanbilge replied:
Cool! Gone!

armanbilge commented:
> I wonder how big the resulting js would be mostly for use on the browser rather than in node, but at any rate this is great work

@ChristopherDavenport did some experiments with this, it seems we might need to do something to reduce the size of the MimeDB which is getting included in the JS file.

rossabaker commented on core/shared/src/main/scala/org/http4s/RangeUnit.scala:
```diff
@@ -0,0 +1,29 @@
+
+import org.http4s.util.{Renderable, Writer}
+
+object RangeUnit {
```
I'm not sure why this is a pure addition.  What is it in 0.23?

armanbilge replied:
https://github.com/http4s/http4s/blob/main/core/src/main/scala/org/http4s/RangeUnit.scala ?

rossabaker replied:
Oh, it showed up as a new file because we lost the attribution in the header, and that was as substantial as the file itself.

rossabaker commented on core/shared/src/main/scala/org/http4s/util/Renderable.scala:
```diff
@@ -51,7 +50,7 @@ object Renderer {
       DateTimeFormatter
         .ofPattern("EEE, dd MMM yyyy HH:mm:ss zzz")
-        .withLocale(Locale.US)
+        // .withLocale(Locale.US) // Causes problems with Scala.js
```
Yikes.  How does scala.js do locales?  I think this is pretty important on machines with Unamerican default locales.

armanbilge replied:
I know how to fix this now! See https://github.com/http4s/http4s/pull/4938#discussion_r683103926

armanbilge replied:
Basically, Scala.js ships with virtually no Locales. Whatever you want, you got to build it yourself, or find someone who has.

rossabaker replied:
Then what is this going to do on a machine whose default language is not English?  We can't have the weekday and month rendered in an arbitrary language and still have valid HTTP dates.

rossabaker replied:
Oh, sorry, I read the grim second comment and not the hopeful comment above it.

armanbilge replied:
Fixed :)

rossabaker commented on dom-core/src/main/scala/org/http4s/dom/package.scala:
```diff
@@ -0,0 +1,41 @@
+  def fromDomHeaders(headers: DomHeaders): Headers =
+    Headers(
+      headers.toIterable.map { header =>
+        header(0) -> header(1)
```
These indexes are freaking me out, but I guess it works?

armanbilge replied:
It really does! 😆

rossabaker commented on dom-fetch-client/src/main/scala/org/http4s/dom/FetchClient.scala:
```diff
@@ -0,0 +1,81 @@
+import fs2.Stream
+import org.http4s.client.Client
+import org.scalajs.dom.crypto._
+import org.scalajs.dom.experimental.Fetch
+import org.scalajs.dom.experimental.Headers
+import org.scalajs.dom.experimental.HttpMethod
+import org.scalajs.dom.experimental.ReadableStream
+import org.scalajs.dom.experimental.RequestInit
```
It's not 1.0 yet, but I wonder what we're getting into as we move toward 1.0 on top of something experimental.  We'll probably have to flag it loudly in docs and release notes?

armanbilge replied:
This is no longer experimental, just binary compatibility. The `scala-js-dom` library is releasing a 2.0 soon with a proper Scala 3 artifact and these moved out of experimental. See https://github.com/scala-js/scala-js-dom/issues/451#issuecomment-891141873

rossabaker reviewed (commented):
Alright, I'm only about halfway through it, but comment density is dropping, so I probably should pause here and pick it up tomorrow.

armanbilge commented on build.sbt:
```diff
@@ -2,29 +2,122 @@ import com.typesafe.tools.mima.core._
+}
+
+lazy val useJSEnv =
+  settingKey[JSEnv]("Use Node.js or a headless browser for running Scala.js tests")
+Global / useJSEnv := NodeJS
+
+ThisBuild / Test / jsEnv := {
+  val old = (Test / jsEnv).value
+
+  val config = SeleniumJSEnv
+    .Config()
+    .withMaterializeInServer("target/selenium", "http://localhost:8889/target/selenium/")
+
+  useJSEnv.value match {
+    case NodeJS => old
+    case Firefox =>
+      val profile = new FirefoxProfile()
+      profile.setPreference("devtools.serviceWorkers.testing.enabled", true)
+      val options = new FirefoxOptions()
+      options.setProfile(profile)
+      options.addArguments("-headless")
+      new SeleniumJSEnv(options, config)
+    case Chrome =>
+      val options = new ChromeOptions()
+      options.setHeadless(true)
+      options.addArguments("--allow-file-access-from-files")
+      val factory = new DriverFactory {
+        val defaultFactory = SeleniumJSEnv.Config().driverFactory
+        def newInstance(capabilities: org.openqa.selenium.Capabilities): WebDriver = {
+          val driver = defaultFactory.newInstance(capabilities).asInstanceOf[ChromeDriver]
+          driver.manage().timeouts().pageLoadTimeout(1, TimeUnit.HOURS)
+          driver.manage().timeouts().setScriptTimeout(1, TimeUnit.HOURS)
+          driver
+        }
+        def registerDriverProvider(provider: DriverProvider): Unit =
+          defaultFactory.registerDriverProvider(provider)
+      }
+      new SeleniumJSEnv(options, config.withDriverFactory(factory))
+  }
+}
```
+1. This is duplicated from the CE build actually 😅

rossabaker replied:
Alright, we'll work that into a plugin somewhere eventually.  Doesn't need to be now.

armanbilge commented on build.sbt:
```diff
@@ -63,8 +163,13 @@ lazy val modules: List[ProjectReference] = List(
   blazeServer,
   blazeClient,
   asyncHttpClient,
+  domCore,
+  domFetchClient,
+  domServiceWorker,
+  domServiceWorkerTests,
```
Tough questions! One way to think about it is these are "native" implementations for their respective platforms, as are the `node-serverless` module and the `ember` modules i.e. requiring no external dependencies for their implementation. If I understand correctly every other module requires additional dependencies. (Technically I'm including `fs2.io` as a "native" dependency here ;)

rossabaker replied:
Okay, let's leave it here for now.  We need a lot of this build complexity anyway for things like the core module.

rossabaker reviewed (commented):
I'm also a bit worried about how large the merge conflicts have been already, and how big they may continue to be as we maintain the production versions.  Would it make sense to spin off the `Platform` traits earlier, on series/0.22 and merge those forward, just to keep the branches in greater sync?  Pay up front now, instead of paying repeatedly later?

armanbilge commented on .gitignore:
```diff
@@ -32,3 +32,5 @@ project/.gnupg
 out
 .vscode/
+
+yarn.lock
```
We had a discussion about it on Discord today, fs2 has the same problem. @mpilquist and @ChristopherDavenport are for removing it. @mpilquist pointed out that it seems to be changing by itself which means what should be a release build could suddenly become a snapshot. It's probably tooling issues.

In our case, I don't think it's a _huge_ deal, this only showed up because I added a web socket client dependency to test out @RaasAhsan's PR on Node.js. We have no such dependencies in the Compile scope.

rossabaker replied:
I guess that's not too severe in this limited case, but it sounds highly problematic in a general case.  I don't see how builds can be repeatable without sharing the lock file.

Is it changing between compile and test?  Is it because of this?  https://github.com/scalacenter/scalajs-bundler/issues/259.

armanbilge replied:
That's a very good point about repeatable builds and thanks for finding that issue, maybe that's all we need to do to fix. I'm curious if this sways @ChristopherDavenport or @mpilquist.

armanbilge commented:
>  Would it make sense to spin off the Platform traits earlier, on series/0.22 and merge those forward, just to keep the branches in greater sync

I'm loathe to do yet another round of these `Platform`, yet that's exactly your point so I have no choice but to agree :) I fixed today's merge conflicts by just starting from scratch for the affected files. No fun.

rossabaker commented:
We don't want to do bidirectional merges.  We could abruptly reverse the direction by merging main to series/0.23 with a strategy of keeping 0.23 to reset the merge base.  But then any development on new features, which is where a lot of the action would happen, would result in "deleted by us" merge conflicts going backward.  Another alternative is to effectively freeze those older branches and selectively cherry-pick back, but the more selective we are, the harder each one gets.  I think merging forward is The Way for as long as we're supporting multiple versions.

You're absolutely right about speculatively splitting the platform on 0.22 not uncovering JVMisms until a couple branches later.  I guess let's stick with the status quo and see how badly it hurts and we can always recalibrate.

rossabaker commented on server/js/src/main/scala/org/http4s/server/middleware/CSRFPlatform.scala:
```diff
@@ -0,0 +1,158 @@
+      }
+    }.map(_.asInstanceOf[SecretKey])
+
+  private[middleware] def digestIsEqual(digesta: Array[Byte], digestb: Array[Byte]) =
```
What is platform specific about this?

armanbilge replied:
Because JVM just has to do `MessageDigest.isEqual(out, decoded)`

rossabaker commented on server/jvm/src/main/scala/org/http4s/server/ServerBuilderPlatform.scala:
```diff
@@ -0,0 +1,32 @@
+
+import java.net.InetAddress
+
+private[server] trait ServerBuilderCompanionPlatform {
```
These were removed in 0.23.

rossabaker commented on server/jvm/src/main/scala/org/http4s/server/defaultsplatform.scala:
```diff
@@ -0,0 +1,47 @@
+
+import java.net.{InetAddress, InetSocketAddress}
+
+private[server] trait defaultsplatform { self: defaults.type =>
```
Nitpicky capitalization

armanbilge commented on ember-client/shared/src/main/scala/org/http4s/ember/client/EmberConnection.scala:
```diff
@@ -42,4 +47,10 @@ private[ember] object EmberConnection {
+  @nowarn("msg=never used")
+  private implicit final class SocketOpsJS[F[_]](val socket: Socket[F]) extends AnyVal {
+    def endOfInput(implicit F: Monad[F]): F[Unit] = F.unit
+  }
```
It could, I wasn't sure what's the right thing to do here. Node.js doesn't support this method and so not making it available is the clearest way to indicate that. I'll raise it up again.

armanbilge commented on laws/src/main/scala/org/http4s/laws/discipline/Ip4sArbitraryInstances.scala:
```diff
@@ -0,0 +1,140 @@
+
+// Unforunately duplicated here from ip4s
+// The provided test-kit requires module support not available in the browser
+private[discipline] trait Ip4sArbitraryInstances {
```
Oops, didn't realize these are published. I'll fix that.

armanbilge commented on node-serverless/src/main/scala/org/http4s/node/serverless/IncomingMessage.scala:
```diff
@@ -14,12 +14,15 @@
-import cats.effect.std.Dispatcher
+import fs2.io.Readable
 
-trait EqF {
```
I think it was unnecessary so I removed it.

armanbilge commented on project/CentralRequirements.scala:
```diff
@@ -14,29 +14,50 @@ object CentralRequirementsPlugin extends AutoPlugin {
-      Developer("rossabaker"           , "Ross A. Baker"         , "ross@rossabaker.com"              , url("https://github.com/rossabaker")),
-      Developer("shengc"               , "Sheng Chen"            , ""                                 , url("https://github.com/shengc")),
-      Developer("SystemFw"             , "Fabio Labella"         , ""                                 , url("https://github.com/SystemFw")),
+      Developer("aeons", "Bjørn Madsen", "bm@aeons.dk", url("https://github.com/aeons")),
```
Yeah this ended up being terrible, my fault for running format on SBT (I think it uses your default .scalafmt config). I'll roll these back.

armanbilge commented on project/Http4sPlugin.scala:
```diff
@@ -292,11 +299,11 @@ object Http4sPlugin extends AutoPlugin {
     val catsEffect = "3.2.1"
     val catsParse = "0.3.4"
-    val circe = "0.14.1"
+    val circe = "0.15.0-M1"
```
I had to cross-build the jawn module upstream, only available in that milestone at the moment.

armanbilge commented on scripts/static_server.py:
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env python3
```
Yes, it runs a static file server for the browser-based tests. I did look into Node and I can do so again, I think it just didn't offer the out-of-the-box static file server but then I ended up customizing the Python one anyway (originally it was just`python3 -m http.server`).

rossabaker replied:
I don't know how many people are going to run the browser tests locally.  Node would be better because it's already necessary elsewhere, but once I enhance the nix-shell, _I_ won't care.

armanbilge replied:
Ok, I investigated this again. The [Node.js docs](https://nodejs.org/en/knowledge/HTTP/servers/how-to-serve-static-files/) literally point you to a 3rd-party library for a static file server.

So I think the options here are:
1. Keep using the python3 static server
2. Add an `npm` dependency to `node-static`
3. Try to implement a simple static file server in pure Node.js
4. Implement a Scala/Java static file server as part of the sbt build

Besides sticking with (1), (4) seems the most feasible. Thoughts?

rossabaker replied:
1 is fine for now.  Most people either aren't going to run the non-JVM tests, or aren't going to be as stubborn as me when asked to globally install a new interpreter for a language as popular as Python.  We can look at 4 as a separate PR, or I can stop whining and improve the Nix shell.

armanbilge replied:
If/when (4) happens, the scaffold server could be rolled into it too. The advantage of having these in the sbt build is they can be more easily started/stopped, currently these scripts have to be started manually before running tests, hope that's okay 😬

armanbilge commented on scripts/scaffold_server.js:
```diff
@@ -0,0 +1,69 @@
+#!/usr/bin/env node
```
Yes, the ClientRouteBatterySuite uses it. I know, not ideal :(

rossabaker replied:
This one is less troublesome than Python because it doesn't add marginally to a full development environment.  I wouldn't sweat this one so much.

armanbilge commented on server/js/src/main/scala/org/http4s/server/SecureSession.scala:
```diff
@@ -0,0 +1,22 @@
+
+import scodec.bits.ByteVector
+
+// TODO not sure about this representation
```
If there's demand, and somebody knows how to parse that raw binary encoding into something intelligble. Unfortunately Node.js just gives us that.

rossabaker replied:
The representation is immutable, but not extensible.  As long as it's unlikely to be extended, what you've done is fine.  Otherwise, we want one of those "contraband" encodings, like the builders.

armanbilge replied:
Actually since I wrote this, I ended up implementing a similar wrapper in fs2.io. Maybe we can use that instead?
https://github.com/typelevel/fs2/pull/2510

armanbilge commented:
Just wanna say a quick thank you to everyone for getting in reviews so quickly and also to @cquiroz for releasing an `en_US` locale lib for Scala.js that we'll use here.

The following is for my own sanity, but let me know if i forgot something:
- [ ] fix locales
- [ ] bring back lockfile
- [ ] fix platform name caps: oauth1platform, defaultsplatform
- [ ] catchNotFatal for crypto
- [ ] make js package internal
- [ ] Add attribution/NOTICE for crypto facade
- [ ] Fix all the headers
- [ ] Drop Http4s.scala
- [ ] bump fs2 and adopt fixes (exceptions, pure socket api, SSLSession class)
- [ ] hide ip4s arbitraries
- [ ] roll back sbt formatting
- [ ] replace Python static file server with Node.js
- [ ] Remove ServerBuilder

armanbilge commented:
@rossabaker I think I've addressed most of your review comments, let me know if I've missed anything. I think I restored all of the headers and I did my best undoing at least some of the build formating. See my comments above re the Node.js/python server scripts and the crypto library.

rossabaker commented on project/Http4sPlugin.scala:
```diff
@@ -67,39 +67,36 @@ object Http4sPlugin extends AutoPlugin {
           attributedSources.contains(baseDirectory.value.toPath.relativize(file.toPath).toString)
+        }
 
         val attributedSources = Set(
```
I wonder whether we even need this hack if sbt-header is leaving the attributions alone now?  That would be lovely, because you just experienced firsthand how fragile this is.

armanbilge replied:
We don't (assuming you are happy with the dual-headers), I can remove it.

rossabaker replied:
Yeah, what you did looks great.  People have tried to tear this project down for stealing work over the years, but I don't think many projects go to our lengths to credit their sources.  Thanks for persevering on this to keep that tradition alive.

armanbilge replied:
I had no idea about that history, really sad to hear because it's readily apparent you've gone above and beyond here.

rossabaker commented:
I'd say privatize whatever you see fit in the JS package, and then we'll land this and cut a milestone.

armanbilge commented:
Hooray!! I'll take a careful pass through the JS package and let you know when it's ready.

> then we'll land this and cut a milestone.

Do you still want to wait for a proper scala 3 artifact for `scala-js-dom` before the milestone?

armanbilge commented:
Actually, I have an idea: `scala-js-dom` (where I got the JS crypto stuff from) is hopefully releasing their 2.0 with Scala 3 artifacts very soon. Maybe I can get in a PR to fix the issue that prevented using their facades in Node.js and forced me to copy their code into http4s. That would be a huge win.

rossabaker commented:
I don't want to keep this PR open longer than necessary.  I would exclude the DOM packages from publishing on Scala 3 for this round, and we can reintroduce later.

I don't think it's terrible if that DOM dependency in core is JS only, if it's necessary to make the core run on JS?  Maybe I'm misunderstanding.

rossabaker commented on core/js/src/main/scala/org/http4s/js/crypto/Crypto.scala:
```diff
@@ -0,0 +1,945 @@
+
+  var iterations: Double = js.native
+
+  var hash: HashAlgorithmIdentifier = js.native
```
Getting a ton of warnings like this.  Maybe we're stuck with them:

```scala
[warn] /Users/ross.baker/src/http4s/core/js/src/main/scala/org/http4s/js/crypto/Crypto.scala:863:42: dead code following this construct
[warn]   var hash: HashAlgorithmIdentifier = js.native
```

armanbilge replied:
I think 2.13 is better with this, because I tried `@nowarn`ing but then 2.13 started complaining there was nothing to `@nowarn`. This could probably be solved with a magic incantation of sbt settings.

rossabaker replied:
I'm not terribly worried about it, but there's enough scala.JS only code, it would be a nice mopup later so we don't miss bugs there by tolerating warnings.

armanbilge commented:
> I don't want to keep this PR open longer than necessary. I would exclude the DOM packages from publishing on Scala 3 for this round, and we can reintroduce later.

Me neither, let's close this out. So it's okay if we do a breaking bump for the `scala-js-dom` from 1.0 to 2.0 for the next milestone?

> I don't think it's terrible if that DOM dependency in core is JS only, if it's necessary to make the core run on JS? Maybe I'm misunderstanding.

Yes, just talking JS here. A DOM (browser) dependency for core is weird because core also runs on non-browser environments such as Node.js. So it should be agnostic.

rossabaker commented:
I took the liberty of drafting a changelog here.  My JavaScript and frontend knowledge is about 15 years out of date, so please don't hesitate to correct anything silly that I said.

> So it's okay if we do a breaking bump for scala-js-dom from 1.0 to 2.0 for the next milestone?

For sure.  1.0 is very much under development.

> A DOM (browser) dependency for core is weird because core also runs on non-browser environments such as Node.js. So it should be agnostic.

And it gets pulled into http4s-core because it provides the JS crypto, otherwise it would be in http4s-dom-core?  Seems like maybe that scala-js-dom is missing some modularity?

armanbilge commented:
> I took the liberty of drafting a changelog here. My JavaScript and frontend knowledge is about 15 years out of date, so please don't hesitate to correct anything silly that I said.

I took a brief look, looks good! 😄

> For sure. 1.0 is very much under development.

Then FWIW, there's no harm to release Scala 3 artifacts using `for3use2_13` for scala-js-dom, because there will _never_ be a `scala-js-dom` v1.x with a Scala 3 artifact. When we get the Scala 3 artifact we will also be bumping up to 2.0 anyway. Totally up to you.

> Seems like maybe that scala-js-dom is missing some modularity

I'd say it's a Scala.js problem as a whole, because they are trying to use a single suffix `_sjs` to support two _very_ different platforms (browsers and Node.js).

I get the feeling your itching to release, say tomorrow? I'll wrap up everything that's left tonight.

rossabaker commented:
> there's no harm to release Scala 3 artifacts using for3use2_13 for scala-js-dom, because there will never be a scala-js-dom v1.x with a Scala 3 artifact.

"Never" is a strong word, and I'd rather not normalize `for3use2_13` in published dependencies.  We'll have a proper fix soon enough.

> I'd say it's a Scala.js problem as a whole, because they are trying to use a single artifact suffix _sjs to support two very different platforms (browsers and Node.js).

I'm not sure the solution to that problem is copying massive amounts of code.  What is the real downside to having all this DOM support in an http4s-core for Node.js?  I don't have a good grasp of what the optimizing compiler does for us here.

> I get the feeling your itching to release, say tomorrow? I'll wrap up everything that's left tonight.

Would love to merge at least.  Release would be great.  It would be a bit cleaner if we had an fs2 milestone to keep like qualifiers with our dependencies.  But not strictly necessary, I guess.  /cc @mpilquist

armanbilge commented:
> "Never" is a strong word

I can say so with certainty because of https://github.com/scala-js/scala-js-dom/issues/451. Basically, the `for3use2_13` has been the designated version for Scala 3, which is why they are forced to do a major version bump to release a _real_ Scala 3 artifact. But, I agree better not to normalize this practice!!

>  What is the real downside to having all this DOM support in an http4s-core for Node.js?

The risk is that that library evolves in ways that strongly diverge from our needs. For example, attempting to use their crypto facade on a non-DOM environment i.e. Node.js is currently throwing a fatal exception, because of a package object initializer that assumes DOM stuff. This is why I copy-pasted the code in the first place. Theoretically, we can try and patch these (and pray that they accept our PRs) but at the end of the day our goals are just not well-aligned.

After crossing many many projects, I do think we need a cross-platform crypto library that supports browsers, Node.js, and maybe even JVM. Http4s, Skunk, and scodec all have this common need.

> Would love to merge at least. Release would be great. It would be a bit cleaner if we had an fs2 milestone to keep like qualifiers with our dependencies.

Yes, let's merge today please. I would also prefer to wait for fs2 to release, but I think that's still be going to be a little while.

Last night and one more time this morning I took a pass through the JS cryptography facade. Unfortunately, I don't think we can make much of it private without being surgical. `CryptoKey` is the essential public interface there, but for users to _obtain_ a `CryptoKey` we have to provide the algorithms and the `SubtleCrypto` API. Hope that's okay.

I just pushed a fix for the build but otherwise I'm 👍 to merge. Thanks!

mpilquist commented:
Hoping for fs2 3.1.0 release tonight

rossabaker commented:
Okay.  I can start a release train this afternoon then, since this will come third.  

Before that happens, someone needs to explain the security alerts to me like I'm a dumb backend developer: #5031
